### PR TITLE
Use specific error message for wrong arity in async builder

### PR DESCRIPTION
**Pull request by jcouv**
Test plan: https://github.com/dotnet/roslyn/issues/51999

Corresponding spec change: https://github.com/dotnet/csharplang/pull/4903
FYI @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks for taking my feedback :smile:

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs [hunk 1]
```diff
@@ -345,10 +345,20 @@ private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F,
                  !builderType.IsVoidType() &&
                  (forMethodLevelBuilder || builderType.DeclaredAccessibility == desiredAccessibility))
             {
-                bool isArityOk = isGeneric
-                                 ? builderType.IsUnboundGenericType && builderType.ContainingType?.IsGenericType != true && builderType.Arity == 1
-                                 : !builderType.IsGenericType;
-                if (isArityOk)
+                if (isGeneric)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:348]: Could you please clarify the difference between `isGeneric` and `builderType.IsGenericType`?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:348]: Is it something like: `isGeneric` refers to whether the task-like thing is generic, while `builderType.IsGenericType` refers to whether the builder is generic?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:348]: Correct. Whether the return type is generic or not drives the rest of the process.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**